Improve backtrace formating while panicking.

Yamakaky opened the pull request:
Fixes #37783.

Done:

- Fix alignment of file paths for better readability
- `RUST_BACKTRACE=full` prints all the informations (current behaviour)
- `RUST_BACKTRACE=(short|yes)` is the default and does:
  - Skip irrelevant frames at the beginning and the end
  - Remove function address
  - Remove the current directory from the absolute paths
  - Remove `::hfabe6541873` at the end of the symbols
- `RUST_BACKTRACE=(0|no)` disables the backtrace.
- `RUST_BACKTRACE=<everything else>` is equivalent to `short` for
  backward compatibility.
- doc
- More uniform printing across platforms.

Removed, TODO in a new PR:

- Remove path prefix for libraries and libstd

Example of short backtrace:
```rust
fn fail() {
    panic!();
}

fn main() {
    let closure = || fail();
    closure();
}
```
Short:
```
thread 'main' panicked at 'explicit panic', t.rs:2
Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: t::fail
            at ./t.rs:2
   1: t::main::{{closure}}
            at ./t.rs:6
   2: t::main
            at ./t.rs:7
```
Full:
```
thread 'main' panicked at 'This function never returns!', t.rs:2
stack backtrace:
   0:     0x558ddf666478 - std::sys::imp::backtrace::tracing::imp::unwind_backtrace::hec84c9dd8389cc5d
                               at /home/yamakaky/dev/rust/rust/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1:     0x558ddf65d90e - std::sys_common::backtrace::_print::hfa25f8b31f4b4353
                               at /home/yamakaky/dev/rust/rust/src/libstd/sys_common/backtrace.rs:71
   2:     0x558ddf65cb5e - std::sys_common::backtrace::print::h9b711e11ac3ba805
                               at /home/yamakaky/dev/rust/rust/src/libstd/sys_common/backtrace.rs:60
   3:     0x558ddf66796e - std::panicking::default_hook::{{closure}}::h736d216e74748044
                               at /home/yamakaky/dev/rust/rust/src/libstd/panicking.rs:355
   4:     0x558ddf66743c - std::panicking::default_hook::h16baff397e46ea10
                               at /home/yamakaky/dev/rust/rust/src/libstd/panicking.rs:371
   5:     0x558ddf6682bc - std::panicking::rust_panic_with_hook::h6d5a9bb4eca42c80
                               at /home/yamakaky/dev/rust/rust/src/libstd/panicking.rs:559
   6:     0x558ddf64ea93 - std::panicking::begin_panic::h17dc549df2f10b99
                               at /home/yamakaky/dev/rust/rust/src/libstd/panicking.rs:521
   7:     0x558ddf64ec42 - t::diverges::he6bc43fc925905f5
                               at /tmp/p/t.rs:2
   8:     0x558ddf64ec5a - t::main::h0ffc20356b8a69c0
                               at /tmp/p/t.rs:6
   9:     0x558ddf6687f5 - core::ops::FnOnce::call_once::hce41f19c0db56f93
  10:     0x558ddf667cde - std::panicking::try::do_call::hd4c8c97efb4291df
                               at /home/yamakaky/dev/rust/rust/src/libstd/panicking.rs:464
  11:     0x558ddf698d77 - __rust_try
  12:     0x558ddf698c57 - __rust_maybe_catch_panic
                               at /home/yamakaky/dev/rust/rust/src/libpanic_unwind/lib.rs:98
  13:     0x558ddf667adb - std::panicking::try::h2c56ed2a59ec1d12
                               at /home/yamakaky/dev/rust/rust/src/libstd/panicking.rs:440
  14:     0x558ddf66cc9a - std::panic::catch_unwind::h390834e0251cc9af
                               at /home/yamakaky/dev/rust/rust/src/libstd/panic.rs:361
  15:     0x558ddf6809ee - std::rt::lang_start::hb73087428e233982
                               at /home/yamakaky/dev/rust/rust/src/libstd/rt.rs:57
  16:     0x558ddf64ec92 - main
  17:     0x7fecb869e290 - __libc_start_main
  18:     0x558ddf64e8b9 - _start
  19:                0x0 - <unknown>
```

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @brson (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

Yamakaky commented:
Any idea about how I could get access to the build root path?

Yamakaky commented:
Result formatting:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ()', src/libcore/result.rs:837
stack backtrace:
   1:    std::sys::imp::backtrace::tracing::imp::write::hfc2edb670e5eda97
   2:    std::panicking::default_hook::{{closure}}::hc66a547fab0b4d38
   3:    std::panicking::default_hook::h7fba1b2f69474bdc
   4:    std::panicking::rust_panic_with_hook::h5d3597668c9f0035
   5:    std::panicking::begin_panic::hfaa38fdbc7d103f6
   6:    std::panicking::begin_panic_fmt::h69e0397ed8fc5362
   7:    rust_begin_unwind
   8:    core::panicking::panic_fmt::h2ba266031787cf7b
   9:    core::result::unwrap_failed::h0a78b8b4914ea99c
             at /home/yamakaky/dev/rust/rust/src/libcore/macros.rs:29
  10:    <core::result::Result<T, E>>::unwrap::h2f7c43c0e2d4b7cd
             at /home/yamakaky/dev/rust/rust/src/libcore/result.rs:737
  11:    t::main::h86d9b74d94eb7fff
             at /tmp/pote/t.rs:2
  12:    __rust_maybe_catch_panic
  13:    std::rt::lang_start::ha33d35ddc5bfd7d2
  14:    main
  15:    __libc_start_main
  16:    _start
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ()', src/libcore/result.rs:837
stack backtrace:
   1:     0x56381e42288e - std::sys::imp::backtrace::tracing::imp::write::hfc2edb670e5eda97
   2:     0x56381e424b91 - std::panicking::default_hook::{{closure}}::hc66a547fab0b4d38
   3:     0x56381e424776 - std::panicking::default_hook::h7fba1b2f69474bdc
   4:     0x56381e424f78 - std::panicking::rust_panic_with_hook::h5d3597668c9f0035
   5:     0x56381e424e12 - std::panicking::begin_panic::hfaa38fdbc7d103f6
   6:     0x56381e424d50 - std::panicking::begin_panic_fmt::h69e0397ed8fc5362
   7:     0x56381e424cd1 - rust_begin_unwind
   8:     0x56381e46db2f - core::panicking::panic_fmt::h2ba266031787cf7b
   9:     0x56381e41cc55 - core::result::unwrap_failed::h0a78b8b4914ea99c
                               at /home/yamakaky/dev/rust/rust/src/libcore/macros.rs:29
  10:     0x56381e41c99b - <core::result::Result<T, E>>::unwrap::h2f7c43c0e2d4b7cd
                               at /home/yamakaky/dev/rust/rust/src/libcore/result.rs:737
  11:     0x56381e41ccd8 - t::main::h86d9b74d94eb7fff
                               at /tmp/pote/t.rs:2
  12:     0x56381e42f0f6 - __rust_maybe_catch_panic
  13:     0x56381e4255a8 - std::rt::lang_start::ha33d35ddc5bfd7d2
  14:     0x56381e41cd12 - main
  15:     0x7f6dc966c290 - __libc_start_main
  16:     0x56381e41c7f9 - _start
```

Mark-Simulacrum commented:
Those appear to be the same, did you copy the wrong thing in?

Yamakaky commented:
For now they are nearly the same, except the address which is not printed with `short`.
An option would be to first construct a `Vec<{demangled, addr, [files]}>`, then print it. Is it OK to alloc in this place, though?
Also, can I use `std::env::current_dir` in this place?

Yamakaky commented:
In fact, that may not be necessary. We can simply recognized the mangled form of `std::panicking::*`, `rust_begin_unwind`, `main` (14 in the stacktrace above) and `_rust_maybe_catch_main` to determine the range of frames to print, and use a double pass. That should catch most of the unwanted frames.

Yamakaky commented:
With `panic!` and without the `::h234567898765` part:
```
thread 'main' panicked at 'explicit panic', t.rs:2
stack backtrace:
   1:    std::sys::imp::backtrace::tracing::imp::write
   2:    std::panicking::default_hook::{{closure}}
   3:    std::panicking::default_hook
   4:    std::panicking::rust_panic_with_hook
   5:    std::panicking::begin_panic
             at /home/yamakaky/dev/rust/rust/src/libstd/panicking.rs:417
   6:    t::main
             at /tmp/pote/t.rs:2
   7:    __rust_maybe_catch_panic
   8:    std::rt::lang_start
   9:    main
  10:    __libc_start_main
  11:    _start
```

estebank commented on src/libstd/sys_common/gnu/libbacktrace.rs:
```diff
@@ -173,7 +178,7 @@ pub fn print(w: &mut Write, idx: isize, addr: *mut libc::c_void,
             if file.is_null() { continue; } // just to be sure
             let file = unsafe { CStr::from_ptr(file).to_bytes() };
-            output_fileline(w, file, line, i == FILELINE_SIZE - 1)?;
+            output_fileline(w, file, line, i == FILELINE_SIZE - 1, format)?;
```
I believe you should be able to accomplish item 4 in #37783 by doing something like

```rust
let mut print_line = if format == backtrace::PrintFormat::Short { false } else { true };
for (i, &(file, line)) in fileline_buf[..fileline_count].iter().enumerate() {
    if file.is_null() { continue; } // just to be sure
    let file = unsafe { CStr::from_ptr(file).to_bytes() };
    if format == backtrace::PrintFormat::Short && !print_line {
        let magic_str = "core::panicking::panic_fmt";
        if file[..magic_str.len()] == magic_str.as_bytes() {
            print_line = true
        }
        continue;
    }
    output_fileline(w, file, line, i == FILELINE_SIZE - 1, format)?;
}
```

Yamakaky replied:
No, this loop only prints the `at file:line` lines.

Yamakaky commented:
It's getting better!
```
thread 'main' panicked at 'explicit panic', t.rs:2
stack backtrace:
{ skipping 5 frames }
   6:    t::test
             at /tmp/pote/t.rs:2
   7:    t::main
             at /tmp/pote/t.rs:5
{ skipping 6 frames }
```
compared to
```
thread 'main' panicked at 'explicit panic', t.rs:2
stack backtrace:
   1:     0x559456964347 - std::sys::imp::backtrace::tracing::imp::write::hd29ea4b7d57abadd
   2:     0x559456969781 - std::panicking::default_hook::{{closure}}::h7cb71ec6c6050f50
   3:     0x559456968376 - std::panicking::default_hook::h7fba1b2f69474bdc
   4:     0x559456968978 - std::panicking::rust_panic_with_hook::h1b9f3a413c4c29f3
   5:     0x559456960263 - std::panicking::begin_panic::h5f62bd622a6e3525
                               at /home/yamakaky/dev/rust/rust/src/libstd/panicking.rs:417
   6:     0x559456960412 - t::test::hc6402ffe28ce1338
                               at /tmp/pote/t.rs:2
   7:     0x55945696042c - t::main::h86d9b74d94eb7fff
                               at /tmp/pote/t.rs:5
   8:     0x5594569710d6 - __rust_maybe_catch_panic
   9:     0x559456968fa8 - std::rt::lang_start::ha33d35ddc5bfd7d2
  10:     0x559456960472 - main
  11:     0x7f05f168c290 - __libc_start_main
  12:     0x559456960089 - _start
```

Yamakaky commented:
Added removal of path prefix. It prints `./<path>` for files in the current directory, `at builtin file <file>` for std and such, and `at library file <file>` for libraries. For now, the path are hardcoded for my system. I need to get the buildroot path, and the paths from `-L`. Any idea?
```
thread 'main' panicked at 'explicit panic', t.rs:2
stack backtrace:
{ skipping 5 frames }
   6:    t::test
             at ./t.rs:2
   7:    t::main::{{closure}}
             at ./t.rs:6
   8:    <core::result::Result<T, E>>::map
             at builtin file libcore/result.rs:465
   9:    t::main
             at ./t.rs:6
{ skipping 6 frames }
```

Yamakaky commented:
Can someone test on windows?

Yamakaky commented:
Updated first post.
r? @alexcrichton

alexcrichton commented on configure:
```diff
@@ -1877,6 +1877,8 @@ fi
 msg
 copy_if_changed ${CFG_SRC_DIR}${INPUT_MAKEFILE} ./Makefile
+# Export CFG_BUILD_DIR for use in backtrace unwinding.
+sed -i -re 's/^(CFG_BUILD_DIR)/export \1/' config.tmp
```
This seems sort of suspicious, how come this was necessary?

Yamakaky replied:
See https://github.com/rust-lang/rust/pull/38165#issuecomment-266630445

alexcrichton commented on src/libstd/sys_common/backtrace.rs:
```diff
@@ -25,45 +27,107 @@ pub const HEX_WIDTH: usize = 18;
     let val = match env::var_os("RUST_BACKTRACE") {
-        Some(x) => if &x == "0" { 1 } else { 2 },
-        None => 1,
+        Some(x) => if ["no", "0"].iter().any(|val| &x == *val) {
```
In the interest of being as conservative as possible here, could we leave this as just recognizing `0`?

Yamakaky replied:
Fixed in ddfef15.

alexcrichton commented on src/libstd/sys_common/backtrace.rs:
```diff
@@ -25,45 +27,107 @@ pub const HEX_WIDTH: usize = 18;
+            None
+        } else if &x == "full" {
+            Some(PrintFormat::Full)
+        } else if ["short", "yes", "1"].iter().any(|val| &x == *val)  {
```
Could this just be collapsed with the case below?

Yamakaky replied:
Are you sure? I find it cleaner with the "catch anything" separated.

alexcrichton commented on src/libstd/sys_common/backtrace.rs:
```diff
@@ -25,45 +27,107 @@ pub const HEX_WIDTH: usize = 18;
 
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum PrintFormat {
+    Full = 2,
```
These values seem to be quite subtle, could you be sure to comment as to their purpose here?

Yamakaky replied:
OK. In fact, it controls if the unwinding should use the short or long backtrace format.

alexcrichton commented on src/libstd/sys_common/backtrace.rs:
```diff
@@ -25,45 +27,107 @@ pub const HEX_WIDTH: usize = 18;
+    let mut already_printed = false;
+    if format == PrintFormat::Short && file_path.is_absolute() {
+        if let Ok(cwd) = env::current_dir() {
+            let buildroot = Path::new(env!("CFG_BUILD_DIR"));
```
Could this use `option_env` and handle the `None` case to ensure that we can build the standard library separately if need be?

Yamakaky replied:
See https://github.com/rust-lang/rust/pull/38165#issuecomment-266630445

Yamakaky replied:
Fixed in 7780ef8

alexcrichton commented on src/libstd/sys_common/backtrace.rs:
```diff
@@ -25,45 +27,107 @@ pub const HEX_WIDTH: usize = 18;
+    let file_path = Path::new(file);
+    let mut already_printed = false;
+    if format == PrintFormat::Short && file_path.is_absolute() {
+        if let Ok(cwd) = env::current_dir() {
```
Most of this can still be printed even if this fails, right?

Yamakaky replied:
Oh yeah, right.

alexcrichton commented on src/libstd/sys_common/backtrace.rs:
```diff
@@ -25,45 +27,107 @@ pub const HEX_WIDTH: usize = 18;
+    if format == PrintFormat::Short && file_path.is_absolute() {
+        if let Ok(cwd) = env::current_dir() {
+            let buildroot = Path::new(env!("CFG_BUILD_DIR"));
+            let libpath = Path::new("/tmp/libpath");
```
Hm I'm curious where this path name came into effect?

Yamakaky replied:
It's a dummy value, because the library directory is defined at compilation of the crate, not of stdlib. I still have to find a way to get it.

alexcrichton commented on src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:
```diff
@@ -85,25 +117,70 @@ pub fn write(w: &mut Write) -> io::Result<()> {
+            if let Some(mangled_symbol_name) = resolve_symname(symaddr) {
+                let magics_begin = [
+                    "_ZN3std3sys3imp9backtrace7tracing3imp5write",
+                    "_ZN3std9panicking",
```
This logic seems like we'd want to do this across all platforms, right? Perhaps this could go into the common module?

Yamakaky replied:
Sure, that would just need a bit of refactoring.

Yamakaky commented:
The `CFG_BUILD_DIR` stuff is a hack until I find a cleaner way to get the build dir. Any idea?
I didn't test on Windows, can someone test?

Yamakaky commented:
Is it possible to also have the appveyor build? It would help me supporting windows.

Yamakaky commented:
@alexcrichton  A solution for cleaner code reuse would be that the unwinding function calls a platform-dependent function that generates some sort of Vec<Frame>, then prints it. Not sure if it's an option, though, since it would allocate while unwinding.

Mark-Simulacrum commented:
The platform specific option could potentially return an iterator instead of a vec to avoid the allocation (through impl Trait, or just explicitly naming it).

Yamakaky commented:
I thought about that, but it may be complicated since the system specific code gives a callback to a C library that's invoked for each frame.

Yamakaky commented:
- Sound like libunwind supports iterative unwinding: http://www.nongnu.org/libunwind/man/unw_step(3).html, but not chosen for rust (see https://github.com/rust-lang/rust/blob/master/src/libstd/sys/unix/backtrace/mod.rs for more background)
- On iOS, the code uses a builtin `backtrace` https://github.com/rust-lang/rust/blob/master/src/libstd/sys/unix/backtrace/tracing/backtrace_fn.rs which sets an array of pointers.
- Windows uses a frame iterator https://github.com/rust-lang/rust/blob/master/src/libstd/sys/windows/backtrace.rs#L142-L156
- The only trouble is gcc. We could convert it to something like the iOS code easily.

What I propose is to add the structs `Backtrace` which implements `Iterator<Frame>`. These two structs would be defined in a system-specific manner, with a common trait for Frame. Only public interface for Frame would be functions to get a CStr symbol name, symbol address and file names and line number. I think it should work, and with only static dispatch. What do you think @alexcrichton?

Yamakaky commented:
@alexcrichton I think I'm ready for a final review. I'm only on linux, so I was not able to test on windows and such. As I moved code around, there is probably at least some unused imports.
I removed the prefix removing for libraries and std as it involves more changes in rustc. That will be for another PR.

alexcrichton commented on src/libstd/sys_common/backtrace.rs:
```diff
@@ -10,60 +10,200 @@
+        output(w, index, *frame, resolve_symname(*frame, &context), format)?;
+        if foreach_symbol_fileline(*frame, |file, line| {
+            output_fileline(w, file, line, format)
+        }, &context)? {
```
The indentation here was a little confusing for me at first, perhaps the result of this could be bound to a `let`?

alexcrichton commented on src/libstd/sys_common/backtrace.rs:
```diff
@@ -10,60 +10,200 @@
+        } else if &x == "full" {
+            Some(PrintFormat::Full)
+        } else if ["short", "yes", "1"].iter().any(|val| &x == *val)  {
+            Some(PrintFormat::Short)
```
I believe we can just remove this, right? (always fall through below)

Yamakaky commented on src/libstd/sys/unix/backtrace/printing/dladdr.rs:
```diff
@@ -32,10 +33,18 @@ pub fn print(w: &mut Write, idx: isize, addr: *mut libc::c_void,
-            CStr::from_ptr(info.dli_sname).to_bytes()
-        }))
+        Some(unsafe {
+            CStr::from_ptr(info.dli_sname).to_str().ok()
```
Oops, the "Some" shouldn't be there

Yamakaky commented on src/libstd/sys/windows/backtrace/printing/msvc.rs:
```diff
@@ -44,30 +53,55 @@ pub fn print(w: &mut Write,
+fn _foreach_symbol_fileline<F>(symbol_addr: Frame,
+                               mut f: F,
+                               context: &BacktraceContext) -> io::Result<bool>
+where F: FnMut(&[u8], libc::c_int) -> io::Result<()>
```
OK

Yamakaky commented on src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:
```diff
@@ -35,28 +36,40 @@ pub fn write(w: &mut Write) -> io::Result<()> {
+        // See libunwind:src/unwind/Backtrace.c for the return values.
+        // No, there is no doc.
+        let ret = match result_unwind {
+            uw::_URC_END_OF_STACK => {
```
http://git.savannah.gnu.org/gitweb/?p=libunwind.git;a=blob;f=src/unwind/Backtrace.c;hb=bc8698fd7ed13a629a8ec3cb2a89bd74f9d8b5c0
`_Unwind_Backtrace` only return some values.

alexcrichton replied:
Sure yeah I can imagine that this function returns a number of values, but we weren't checking any of them before, only for errors. What prompted the change in logic?

Yamakaky replied:
I thought it was better to handle error codes. Was I wrong?

Yamakaky replied:
No, I just found it strange to ignore the error case (https://github.com/rust-lang/rust/blob/master/src/libstd/panicking.rs#L349, I'm watching ^^)

alexcrichton replied:
Hm I feel like we're still not getting to an answer to my question. The code you linked to ignores errors because it has nowhere to pass the errors to and the information is just advisory, so it's fine to just ignore errors.

My question is *why* did this change happen? Is it purely being proactive? Was it reactive to some problem?

Yamakaky replied:
No, I added it for the reasons explained above, without detected problems

Yamakaky commented on src/libstd/sys/windows/backtrace/mod.rs:
```diff
@@ -129,31 +88,84 @@ unsafe fn _write(w: &mut Write) -> io::Result<()> {
         if addr == frame.AddrReturn.Offset || addr == 0 ||
            frame.AddrReturn.Offset == 0 { break }
 
+        frames[i] = addr;
```
Yeah, not sure why there is this `-1`, but it was probably important XD

Yamakaky commented on src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:
```diff
@@ -35,28 +36,40 @@ pub fn write(w: &mut Write) -> io::Result<()> {
+                match cx.last_error {
+                    Some(err) => panic!("Fatal error while unwinding: {}",
+                                        err),
+                    None => panic!("Fatal error while unwinding"),
```
Ok

Yamakaky replied:
What do you suggest as return value?

alexcrichton replied:
Oh a bool or a Result would be fine

Yamakaky replied:
Yeah, sure, now that there is no cx.next_error ^^

Yamakaky replied:
In fact, if I replace it with a bool we will loose the io::Error on https://github.com/Yamakaky/rust/blob/better-backtrace/src/libstd/sys/windows/backtrace/mod.rs#L76. Is that what we want?

Yamakaky replied:
An alternative would be to return the already-decoded frames on error. Are you OK with that?

alexcrichton commented:
cc @rust-lang/libs I think this is getting close to merging, and would be good to get some more eyes on the new output format here!

sfackler commented:
Assuming the example in the PR message is up to date, this looks awesome!

Yamakaky commented:
It is, except the "builtin file" part ;)

Yamakaky commented on src/libstd/sys/windows/backtrace/printing/msvc.rs:
```diff
@@ -0,0 +1,107 @@
+
+/// Converts a pointer to symbol to its string value.
+pub fn resolve_symname(symbol_addr: Frame, context: &BacktraceContext)
+    -> Option<&'static str>
```
Maybe then tie its lifetime to the one of `BacktraceContext`?

alexcrichton replied:
Unfortunately on Windows at least the lifetime is fundamentally tied to the stack frame, so it'll need to take a closure or something like that.

Yamakaky replied:
Do you know where in the doc it's written? At least, it's not at https://msdn.microsoft.com/en-us/library/windows/desktop/ms681323(v=vs.85).aspx. That would be good to add as a comment.

alexcrichton replied:
This is not documented because it's just part of the API. If you take a look at the implementation of `resolve_symname` you'll see that it creates a local `SYMBOL_INFO` on the stack. The function then returns a pointer into this structure. That'll cause segfaults on MSVC when executed.

Yamakaky replied:
Oh, I see. The troubles of `unsafe` ^^
A possibility would be to return a Cow. That way, gnu returns a str and msvc returns a String.
Another would be to use a callback, as you said. The implementation would be a bit more complicated, though.

alexcrichton commented on src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:
```diff
@@ -35,28 +35,29 @@ pub fn write(w: &mut Write) -> io::Result<()> {
             }
-            _ => Ok(()),
+            _ => unreachable!("\"{:?}\" should not be triggered",
+                              result_unwind),
```
Er sorry, what I mean is that this *must* not panic. It should return an error of some kind (`Err(())` is fine) so the caller can deal with the failure.

purpleposeidon commented on src/libstd/panicking.rs:
```diff
@@ -243,8 +247,8 @@ fn default_hook(info: &PanicInfo) {
+            if let Some(format) = log_backtrace {
+                let _ = backtrace::print(err, format);
             } else if FIRST_PANIC.compare_and_swap(true, false, Ordering::SeqCst) {
                 let _ = writeln!(err, "note: Run with `RUST_BACKTRACE=1` for a backtrace.");
```
The help message below should be updated to say something like
    note: Run with `RUST_BACKTRACE=full` or `RUST_BACKTRACE=short` for a backtrace.

Yamakaky replied:
Yes, good idea

estebank replied:
I don't see this message having changed.

purpleposeidon commented:
I'd love me some easy-reading stack traces! I have some hopefully-not-too-bike-shedding remarks.

Here's how I'd format it:
```
thread 'main' panicked at 'explicit panic', t.rs:2
Some details are omitted; run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
    t::test @ t.rs:2
    t::main::{{closure}} @ t.rs:6
    <core::result::Result<T, E>>::map @ std libcore/result.rs:465
    t::main @ t.rs:6
```

The frame numbers, while not particularly distracting, AFAIK aren't generally useful. Same story with `{ skipping N frames }`.

Java stacktraces have the function + symbol position on the same line, which certainly keeps things short, but is maybe a bit too dense:
```
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
```

(It's calling `std` "std" instead of `builtin` now?)

It might be cool to use some <tt>terminal::escape::<b>codes</b></tt> on the function names to help them stand out from the package path. (Technology from the past! Save the future!) Maybe that's taking it too far? Parsing terminfo sounds like too much for a stacktrace printer to do...

Yamakaky commented:
Good ideas. The printing format is a quick change, there are certainly things to improve.
I removed the prefix removal for std for now, it will be for another PR.

Yamakaky commented:
rebased to upstream

alexcrichton commented:
@Yamakaky this still exhibits the unsafety on MSVC I believe I pointed out before, so we can't merge until that's resolved.

Yamakaky commented:
Sorry, forgot about this problem. Should be good now.

oli-obk commented:
travis complains:

> /checkout/src/libstd/sys_common/backtrace.rs:51: line longer than 100 chars
> /checkout/src/libstd/sys/windows/backtrace/printing/mod.rs: incorrect license

Yamakaky commented:
Done

oli-obk commented:
> failures:
>    [run-pass] run-pass/backtrace-debuginfo.rs
>    [run-pass] run-pass/backtrace.rs

Yamakaky commented:
Not sure what's the problem...

oli-obk commented on src/test/run-pass/backtrace-debuginfo.rs:
```diff
@@ -142,7 +142,7 @@ fn run_test(me: &str) {
 
     let mut template = Command::new(me);
-    template.env("RUST_BACKTRACE", "1");
+    template.env("RUST_BACKTRACE", "full");
```
also need to adjust the backtrace a few lines below

Yamakaky replied:
Doesn't solve the problem...

alexcrichton commented on src/libstd/sys/windows/backtrace/printing/msvc.rs:
```diff
@@ -0,0 +1,108 @@
+
+        if ret == c::TRUE {
+            let ptr = info.Name.as_ptr() as *const c_char;
+            CStr::from_ptr(ptr).to_str().ok().map(Cow::Borrowed)
```
This is still a memory safety violation, it's returning a borrowed pointer into a stack-local.

This function should likely just take a closure to call.

Yamakaky replied:
Oops, wrong variant.

alexcrichton commented on src/libstd/sys/windows/backtrace/printing/msvc.rs:
```diff
@@ -0,0 +1,108 @@
+
+        if ret == c::TRUE {
+            let ptr = info.Name.as_ptr() as *const c_char;
+            CStr::from_ptr(ptr).to_str().ok().map(|s| Cow::Owned(s.into()))
```
I think this is still an issue in the sense that we should use a closure here. We should strive to avoid allocations in backtraces because they're often generated in limited context situations (panics, aborts, etc), where touching as little as possible is advantageous.

Yamakaky replied:
Ok, let's use a callback then ^^

Yamakaky commented:
Done.
Where do you think I should put some doc about the functions defined in `sys/`? In the header of `sys_common/backtrace.rs`?

alexcrichton commented:
Implementation-wise this is looking better to me, thanks! I think the locking story still needs to be fixed up, though. Previously everything was basically single-threaded for most implementations because most of the library support here isn't threadsafe (e.g. windows and libbacktrace). In the restructuring locks were preserved for windows but not for libbacktrace (e.g. when resolving symbols), I believe.

Can the locking all be moved to the common path? That way implementations don't have to have lots of little locks themselves.

Yamakaky commented:
You mean, we put a lock to protect `backtrace::print`?

alexcrichton commented:
Yeah just throwing a lock around the whole operation is fine (it's wanted anyway to avoid concurrent backtrace prints)

Yamakaky commented:
Should I remove the lock in the windows implementation then?

alexcrichton commented on src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:
```diff
@@ -35,28 +35,29 @@ pub fn write(w: &mut Write) -> io::Result<()> {
+            uw::_URC_END_OF_STACK | uw::_URC_FATAL_PHASE1_ERROR => (),
+            _ => {
+                println!("Unwinding error: \"{:?}\" should not be triggered",
+                         result_unwind);
```
Hm I think I've mentioned this before, but this cannot do operations such as print or allocate. This needs to return an error code.

alexcrichton commented on src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:
```diff
@@ -35,28 +35,29 @@ pub fn write(w: &mut Write) -> io::Result<()> {
             }
-            _ => Ok(()),
         };
         LOCK.unlock();
```
Locks like this can be removed now that the global operation is locked.

alexcrichton commented on src/libstd/sys_common/backtrace.rs:
```diff
@@ -10,60 +10,224 @@
+pub fn print(w: &mut Write, format: PrintFormat) -> io::Result<()> {
+    static LOCK: Mutex = Mutex::new();
+
+    // Use a lock to prevent mixed output in multithreading context.
```
Can you expand this to say it's also required for memory safety on most platforms?

alexcrichton commented:
@Yamakaky any luck with the failing tests?

Yamakaky commented:
I don't really understand what the tests do so it's difficult to fix them ^^

alexcrichton commented:
Given the [test logs](https://travis-ci.org/rust-lang/rust/jobs/192247508#L1645) it looks like output isn't being found when it's expected. Looking at the test it seems that some of the line number information is slightly off. Are you sure that the method backends are called for information is the exact same as before?

alexcrichton commented:
I think this is close enough that we're ready to consider merging. I'm going to propose merging this but I'm curious how @rust-lang/libs feels about the change in format by default for backtraces.

@rfcbot fcp merge

Yamakaky commented:
Update output example.

brson commented:
It looks to me like the op switches the display order of stack frames, so main is displayed first, whereas it used to be last. Is that accurate? If that's the case, is the 'full' option also reversed, or is it in the 'old' order for backwards compatibility (I think we should be consistent in the order)?

Can the formatting scheme here be duplicated out of tree so that others using libbacktrace can imitate the official formatting (error-chain e.g. captures backtraces)?

Yamakaky commented:
In fact, I think we could even extract the entire backtrace handling to a separate crate. Could we do that after this PR, though? ^^

alexcrichton commented:
@brson I personally expected the ordering to stay the same, just the prefix/suffix removed and the output a little different. @Yamakaky can you confirm that the ordering is the same?

Yamakaky commented:
I don't remember changing the order, I'll check.

Yamakaky commented:
No, it's the same order. Someone talked about reversing the order, but it was not implemented.

emilk commented:
To (optionally) reverse the stack trace order is a separate issue and should/will be a separate PR: https://github.com/rust-lang/rust/issues/38873#issuecomment-271113158

Yamakaky commented:
Yes, that's it ^^

aturon commented:
Looks beautiful!

bors commented:
:umbrella: The latest upstream changes (presumably #39234) made this pull request unmergeable. Please resolve the merge conflicts.

Yamakaky commented:
I think I found why the test fails. The backtrace contains the file:line of the functions and not the file:line of the lines that failed. Just have to find why that's the case.

Yamakaky commented:
Yeay !
I think we could take all the backtrace handling code in std, convert it to `no_std` and use it as a dependency of std. That way, it could be published to crates.io and used as a dependency of https://crates.io/crates/backtrace. Could we do that in an other PR, though?

Yamakaky commented:
Anything missing before merge?
Also, moving the conversation about putting the code in a separate crate to https://github.com/rust-lang/rust/issues/39503.

alexcrichton commented:
ping @brson (checkbox)

Yamakaky commented:
Oh, yeah, right.

brson commented:
@Yamakaky can you paste some output for how the backtraces look now?

Yamakaky commented:
Done in the first post of this thread.

rkruppe commented:
Looks great! I noticed that `main` is omitted from the abridged backtrace, which is a bit surprising to me since main is *not* standard library code or other noise, but is instead (usually) written by the user.

alexcrichton commented:
@Yamakaky should the documentation be updated then? The docs in the book were changed but seem to reflect what I believe at this point is historical syntax.

Yamakaky commented:
Oh yeah, right. What should I change?

alexcrichton commented:
`src/doc/book/functions.md` is the one I'm looking at, and I think it may be the only one.

Looking at the current format, how come the numbers were elided? I could imagine that without file/line numbers that the format may be difficult to parse if there's not line numbers demarcating entries perhaps?

Can you also extend an existing test (or add a new one) which asserts that by default the standard library pieces aren't printed? And then with `RUST_BACKTRACE=full` they are indeed printed?

brson commented:
Thanks @Yamakaky.

brson commented:
@Yamakaky do you mind showing both the short and full cases for the same stack trace, for comparison?

Yamakaky commented:
Done for the doc
@brson done
On it for the test.
Line numbers: makes sense, on it.

Yamakaky commented:
Done for the frame number

nrc commented:
In the example in the OP, the short and full traces don't seem to correspond? One has `fail` and one has `diverge`.

Is it correct that this only elides frames from the top and bottom of the trace?

Yamakaky commented:
Oh, should check for the doc.
Yes, it doesn't remove frames in the middle.

Yamakaky commented:
Added test

alexcrichton commented on src/test/run-pass/backtrace.rs:
```diff
@@ -59,6 +59,49 @@ fn runtest(me: &str) {
+    let should_be_present = &[
+        "std::panicking",
+        "__rust_maybe_catch_panic",
+        "__libc_start_main",
```
Asserting this symbol and `_start` below are present is likely to fail on non-linux platforms, perhaps those symbols could be tested on just Linux? Otherwise looks good to me!

Yamakaky replied:
Done

alexcrichton commented:
Thanks for the update @Yamakaky as well as the gists in the OP, the outputs look great to me.

@brson, how do you feel about those traces?

Yamakaky commented:
Done

brson commented:
@Yamakaky thanks for the updates. The results look fantastic.

The only qualm I have is slight:

```
thread 'main' panicked at 'explicit panic', t.rs:2
Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
```

Here there are three lines of text output and the capitalization and punctuation appear a bit inconsistent. Particularly with that capital `Some`.

Here's what happens today on a non-backtrace panic:

```
thread 'main' panicked at 'explicit panic', test.rs:2
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

So istm there are three modes here, no backtrace, short backtrace, full backtrace, and they should be as consistent as possible. For sake of comparison here's the full case as of this PR:

```
thread 'main' panicked at 'This function never returns!', t.rs:2
stack backtrace:
```

Here's an attempt to make them more consistent:

```
thread 'main' panicked at 'explicit panic', t.rs:2
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
```

```
thread 'main' panicked at 'explicit panic', test.rs:2
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

```
thread 'main' panicked at 'This function never returns!', t.rs:2
stack backtrace:
```

Yes, that looks good to me. The only thing I changed was to preface 'Some details are omitted' with 'note:'.

@Yamakaky can you make that one change?

I'm happy for this to fcp now. Thanks @Yamakaky, this is a great improvement.

brson commented:
@nrc Yes I believe that is correct - the op examples are from different source.

Yamakaky commented:
I can add 6 chars ;)

alexcrichton commented:
Great! @Yamakaky can you squash the commits down and I'll r+?

Yamakaky commented:
You mean all in one commit?

alexcrichton commented:
Multiple commits are fine (if they're logically separated), but this probably doesn't need 54 commits

Yamakaky commented:
Should be good.

alexcrichton commented:
@bors: r+

Thanks again for your patience @Yamakaky!

bors commented:
:pushpin: Commit 6277694 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 62776946e476bcb09e9a93330d862384be47c4f2 -->

bors commented:
:umbrella: The latest upstream changes (presumably #39638) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:lock: Merge conflict

Yamakaky commented:
Done

alexcrichton commented:
Looks like Travis failed?

Yamakaky commented:
Oops, merge problem, should be good this time.

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit fdd8c34 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton fdd8c349ef5314d286a6227518cbda0036dfff3d -->

bors commented:
:hourglass: Testing commit fdd8c34 with merge e1045f4...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1866)

frewsxcv commented:
@bors r-

https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1865/job/5rbccsyx4beo1jqg

Yamakaky commented:
Missing comma in the mvsc code.

frewsxcv commented:
@bors r=alexcrichton

bors commented:
:pushpin: Commit 54ffbe8 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 54ffbe81dc37194e4b631aa6e747f5ff390b2315 -->

bors commented:
:hourglass: Testing commit 54ffbe8 with merge d9affc4...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1868)

Yamakaky commented:
We will see all the compilations errors I didn't see on Linux ^^

Yamakaky commented:
Will it be good this time?

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 86f5709 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 86f5709851e480da92fffc1e17fbd8c01ef17f9a -->

bors commented:
:hourglass: Testing commit 86f5709 with merge 9149b16...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1873)

Yamakaky commented:
- Fixed issues
- Add unsafe blocks since windows version of `unwind_backtrace` is no longer unsafe.

Yamakaky commented:
Could you relaunch the build please?

aturon commented:
@bors r=alexcrichton

bors commented:
:pushpin: Commit c9b125b has been approved by `alexcrichton`

<!-- @bors r=alexcrichton c9b125b34d405c12073595e32d3426c5d2e41635 -->

bors commented:
:hourglass: Testing commit c9b125b with merge 20bf23e...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1890)

Yamakaky commented:
Oops, should be `return e` instead of `e?`.

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 67d17e5 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 67d17e5eedc20f6d65b13e6943885e3ee4af1003 -->

bors commented:
:hourglass: Testing commit 67d17e5 with merge 9b323ce...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1894)

Yamakaky commented:
Forgot a `Result?`

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 95ba9ab has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 95ba9ab4d97c725537a2a15a80c7af050ef0737d -->

bors commented:
:hourglass: Testing commit 95ba9ab with merge 9ff8886...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1897)

Yamakaky commented:
We are at the unsafe checks, so we are near the end. Sorry for the multiple build request...

petrochenkov commented:
@Yamakaky 
Maybe you could test it locally before sending to bors, using VMs or cross-compilation?
Even with slow machine it's faster than "fix -> reapprove -> send to Travis/Appveyor -> fail" cycle.

Yamakaky commented:
I'm stupid, I forgot about cross compilation...

petrochenkov commented:
@bors r=alexcrichton
(The queue is empty anyway.)

bors commented:
:pushpin: Commit 170251d has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 170251d2faa02ef53ec01d32963720542f997a49 -->

bors commented:
:hourglass: Testing commit 170251d with merge d4a901b...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1904)

Yamakaky commented:
I have trouble building

petrochenkov commented:
@bors r=alexcrichton

bors commented:
:pushpin: Commit 090e5a3 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 090e5a354830fdc01520ba9ca341c480c418412b -->

bors commented:
:hourglass: Testing commit 090e5a3 with merge 4b3f07d...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/200664353)

Yamakaky commented:
Errors in mac code now XD

petrochenkov commented:
@bors r=alexcrichton

bors commented:
:pushpin: Commit 03cae42 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 03cae429d8eabdd63688fa5f14776e487cfcf078 -->

bors commented:
:hourglass: Testing commit 03cae42 with merge dbc7e2b...

Yamakaky commented:
Sorry, forgot to push a fix

petrochenkov commented:
@bors r-

petrochenkov commented:
@bors r=alexcrichton

bors commented:
:pushpin: Commit 1522faa has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 1522faac188cf9a0dbcab432c2cc77e5588ecbbe -->

bors commented:
:hourglass: Testing commit 1522faa with merge b1b82e1...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1907)

Yamakaky commented:
Sorry again @alexcrichton and @petrochenkov for all the roundtrips...

Yamakaky commented:
This fail doesn't seem to be my fault

petrochenkov commented:
```
appveyor DownloadFile https://api.pub.build.mozilla.org/tooltool/sha512/%SCCACHE_DIGEST%
Error downloading remote file: One or more errors occurred.
Inner Exception: Remote server returned 500: Internal Server Error
Command exited with code 2
```
Looks spurious.
@bors r=alexcrichton

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.

bors commented:
:pushpin: Commit 1522faa has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 1522faac188cf9a0dbcab432c2cc77e5588ecbbe -->

petrochenkov commented:
@bors retry

bors commented:
:hourglass: Testing commit 1522faa with merge fc059a8...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1908)

petrochenkov commented:
```
appveyor DownloadFile https://api.pub.build.mozilla.org/tooltool/sha512/%SCCACHE_DIGEST%
Error downloading remote file: One or more errors occurred.
Inner Exception: Remote server returned 500: Internal Server Error
Command exited with code 2
```
Ok, it doesn't looks spurious anymore.
Let's wait for @alexcrichton.

Yamakaky commented:
^^

alexcrichton commented:
It looks like https://api.pub.build.mozilla.org is down at the moment, so we'll need to wait for it to come back online

alexcrichton commented:
@bors: retry

bors commented:
:hourglass: Testing commit 1522faa with merge 22f651c...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1913)

Yamakaky commented:
The mac build is still failing because of api.pub.

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 7af51f6 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 7af51f684daf02534b07575e87e49a5e212c49e4 -->

bors commented:
:hourglass: Testing commit 7af51f6 with merge ef5fef9...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1916)

Yamakaky commented:
Warning of a field only used in some backend: it smells the end.

petrochenkov commented:
@bors r=alexcrichton

bors commented:
:pushpin: Commit 0f6d17f has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 0f6d17f0c260c2a23ca555d929d34c6c9b61a6e1 -->

bors commented:
:hourglass: Testing commit 0f6d17f with merge ed47c43...

Yamakaky commented:
Again, why isn't the appveyor build enabled for all pull requests?

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/200889390)

Yamakaky commented:
The emscripten build seems broken

alexcrichton commented:
@bors: retry

* sccache wedged #39003

On Sun, Feb 12, 2017 at 10:37 AM, Yamakaky <notifications@github.com> wrote:

> The emscripten build seems broken
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/38165#issuecomment-279230305>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD95AbbBRloAlV3o_24CDgXo8hsiEjmks5rbzVNgaJpZM4LDtfR>
> .
>

bors commented:
:hourglass: Testing commit 0f6d17f with merge a66dc5c...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/200906469)

Yamakaky commented:
Moar syntax error, this time in osx code.

petrochenkov commented:
The ride never ends.jpg
@bors r=alexcrichton

bors commented:
:pushpin: Commit 767bb2a has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 767bb2a17282435bbbcba8c428a53b7df83f37b6 -->

Yamakaky commented:
Builds failure for eternity

bors commented:
:hourglass: Testing commit 767bb2a with merge 6563127...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/200950791)

bors commented:
:umbrella: The latest upstream changes (presumably #39633) made this pull request unmergeable. Please resolve the merge conflicts.

Yamakaky commented:
Fix typo, rebased

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 485ecd7 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 485ecd7cec68065a94e4bf13cbf526274df5dc37 -->

Yamakaky commented:
BTW @alexcrichton: after this PR is merged, please @me in the following issues about backtraces. Now that I know how it works ;)

bors commented:
:hourglass: Testing commit 485ecd7 with merge 2e59ee2...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/201925537)

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit d50e4cc has been approved by `alexcrichton`

<!-- @bors r=alexcrichton d50e4cc0640e54a64d0f7ccb05a77fd4a2fe0741 -->

bors commented:
:hourglass: Testing commit d50e4cc with merge 9b7ac2b...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/202475151)

alexcrichton commented:
@Yamakaky the same tests keeps failing, can you investigate?

Yamakaky commented:
I am!

alexcrichton commented:
Thanks @Yamakaky! Feel free to ping me with questions or whenever this PR is updated.

petrochenkov commented:
@Yamakaky 
I've tested this PR locally and will send you the fix soon.
The errors on Windows are non-trivial, you'd never land this without local debugging.

Yamakaky commented:
Cool, thanks!

petrochenkov commented:
@Yamakaky 
This condition may be too strict:
```
    for symbol in removed_symbols {
        assert!(!s.contains(symbol),
                "{} should be removed from the backtrace",
                symbol);
    }
```
Symbols in the bottom of the stack are very platform specific, mangled name returned from `resolve_symname` is platform specific too, some targets have incomplete backtraces with garbage like `mingw_set_invalid_parameter_handler` or even `<unknown>`.
If anything of this stays unremoved, then `std::panicking` somewhere in the middle of backtrace will stay too, and the test will fail.
I recommend to disable this check for everything except for few tier1 platforms with good backtrace support if Travis continues failing.

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 0982a28 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 0982a2830460e20524d0227f83a21ce4a37c94e5 -->

bors commented:
:hourglass: Testing commit 0982a28 with merge a5c2937...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/204909062)

petrochenkov commented:
run-pass/backtrace.rs fails on ARM in qemu due to this failed assertion
```
    assert!(s.contains("stack backtrace") && s.contains(&expected("foo")),
            "bad output: {}", s);
```
because the output looks like this
```
thread 'main' panicked at 'explicit panic', /checkout/src/test/run-pass/backtrace.rs:24
```
and doesn't contain neither `foo` nor any backtrace at all.

Yamakaky commented:
Maybe, as you said, only test the tier 1 platforms?

petrochenkov commented:
@Yamakaky 
The ARM issue from Travis is a regression, not just a lacking backtrace support.
I've reproduced it user mode qemu as well and it doesn't happen on master.
`_Unwind_Backtrace` returns bad `_Unwind_Reason_Code` with this patch for some reason, I haven't investigated further yet.

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 53a5d56 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 53a5d564354727debf37ae6bd807dd6fd65308dc -->

bors commented:
:hourglass: Testing commit 53a5d56 with merge 957c33c...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/205346291)

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit c8b1742 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov c8b17427c2b4dd3eed005b3a30dcb94892144825 -->

bors commented:
:hourglass: Testing commit c8b1742 with merge 7f82f03...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.2125)

petrochenkov commented:
Some new frame that needs to be filtered away on Windows/GNU, probably due to enabled debug assertions on Appveyor, I tested everything in release mode.

petrochenkov commented:
@Yamakaky 
I think we need to remove the `// Make sure than the short version cleans the backtrace.` part of the test completely.
It was tested manually and works reasonably in common configurations, but supporting the *guarantee* that all these symbols will be filtered on all platforms in all configurations will be a huge PITA, given that we can't do this even for officially tested configurations.

petrochenkov commented:
Meanwhile, I fixed the test in Windows+Debug configurations (https://github.com/Yamakaky/rust/pull/4), let's try to land it as is once more.

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 9e054cf has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 9e054cf59dba2cb6612da745d663ab74be7fb3a7 -->

Yamakaky commented:
So you think this part of the test should be removed?

petrochenkov commented:
@Yamakaky 
If it fails 1-2 more Travis runs, then yes.

bors commented:
:hourglass: Testing commit 9e054cf with merge 4b31fca...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/205585213)

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit d0fe018 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov d0fe018cd452f065f99671f966529e5d5c894aaa -->

bors commented:
:hourglass: Testing commit d0fe018 with merge fe43a3b...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/205627218)

Yamakaky commented:
So close!

petrochenkov commented on src/libstd/sys_common/backtrace.rs:
```diff
@@ -126,7 +126,8 @@ fn filter_frames(frames: &[Frame],
         "ZN3std9panicking",
-        "std::panicking",
+        "ZN3std5panic",
+        "std::panic",
```
You need to add `std::panic` in all variations:
```
"_ZN3std5panic" // Linux
"ZN3std5panic" // Windows/GNU
"std::panic" // Windows/MSVC
```
(I'd also keep `"std::panicking"` just for clarity)

Yamakaky replied:
Oops, right

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 52bed53 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 52bed53e999c8628ccb47d349283be65b42a6dd2 -->

bors commented:
:hourglass: Testing commit 52bed53 with merge bea2b27...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.2140)

petrochenkov commented:
Both Travis and Appveyor failed.
[Travis](https://travis-ci.org/rust-lang/rust/jobs/205718187) on the next test condition
```
'the full version of the backtrace should contain __libc_start_main', /checkout/src/test/run-pass/backtrace.rs:103
```
[Appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.2140/job/bw7p30ayxw7tbrjt) on the "classic" one (WTF, I tested this locally in this exact configuration)
```
thread 'main' panicked at 'explicit panic', C:\projects\rust\src/test\run-pass\backtrace.rs:24
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: <std::time::SystemTimeError as core::fmt::Display>::fmt
   1: std::panicking::Location::line
   2: std::panicking::Location::line
   3: std::panicking::rust_panic_with_hook
   4: std::panicking::begin_panic<&str>
             at ./src\libstd\panicking.rs:511
   5: backtrace::foo
             at ./src\test\run-pass\backtrace.rs:24
   6: backtrace::main
             at ./src\test\run-pass\backtrace.rs:156
'
```

Time to drop both `Make sure than the short...` and `Only on linux for _start...` conditions.

Yamakaky commented:
I'm thinking: what if we captured a stacktrace juste before the main, then substract it to the backtrace on panic? That way, at least the bottom of the trace would be correct all the time. Maybe we could do the same for the top, in `std::panicking::begin_panic`, `core::result::unwrap_failed`... It would be easier to maintain.

petrochenkov commented on src/test/run-pass/backtrace.rs:
```diff
@@ -58,30 +58,6 @@ fn runtest(me: &str) {
-    }
     assert!(s.contains(" 0:"), "the frame number should start at 0");
 
     // Only on linux for _start and __libc_start_main
```
This one failed on Travis too.

petrochenkov commented:
@Yamakaky 
Maybe as a next PR?
The problem of this patch is that it already tries to change far too much cross-platform code at once.
If `main` and panic start can be reliably detected in presence of inlining and custom `main`s, then yes, it would be a more robust solution, this needs some experimentation.

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 6398b20 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 6398b2078d92c583c4c7eb76312d474042a2c68f -->

bors commented:
:hourglass: Testing commit 6398b20 with merge 4be034e...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.2142), [status-travis](https://travis-ci.org/rust-lang/rust/builds/205873439)
Approved by: petrochenkov
Pushing 4be034e62270ca15c9fff173faef11939092f4e3 to master...

Yamakaky commented:
\o/
Finally!

jdm commented:
Does this also affect the output of https://crates.io/crates/backtrace or should we file a separate issue in that repository?

Yamakaky commented:
You should open an issue.
